monitor/top_stats.hpp

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2015+ Budnik Andrey <budnik27@gmail.com>
+ *
+ * This file is part of Elliptics.
+ *
+ * Elliptics is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Elliptics is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Elliptics.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef __DNET_MONITOR_TOP_STATS_HPP
+#define __DNET_MONITOR_TOP_STATS_HPP
+
+#include "event_stats.hpp"
+#include "library/elliptics.h"
+
+/*
+ * Default number of top keys returned by top keys statistics provider
+ */
+#define DNET_DEFAULT_MONITOR_TOP_LENGTH 50
+
+/*
+ * Default limit of memory for collecting information about events, in bytes.
+ * Event approximate size is 100 bytes, so default size should be enough for storing ~10000 events
+ */
+#define DNET_DEFAULT_MONITOR_TOP_EVENTS_SIZE 1000000
+
+/*
+ * Default top keys statistics window size in seconds, 300s = 5 minutes
+ */
+#define DNET_DEFAULT_MONITOR_TOP_PERIOD 300
+
+namespace ioremap { namespace monitor {
+
+class key_stat_event;
+
+}}  /* namespace ioremap::monitor */
+
+namespace ioremap { namespace cache {
+
+template<>
+struct treap_node_traits<ioremap::monitor::key_stat_event>
+{
+	typedef const struct dnet_id* key_type;
+	typedef size_t priority_type;
+};
```

Comment:
ioremap::cache::treap_node_traits used in ioremap::cache::treap_node_t
This is concrete template class specialization for type ioremap::monitor::key_stat_event:
> typedef const struct dnet_id* key_type; // for ioremap::monitor::key_stat_event
> typedef const uint8_t* key_type; // for ioremap::cache::data_t

Different types, so it is not possible to merge it into single data structure.